src/testing_util/openssl_assertions.h

```diff
@@ -0,0 +1,138 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef KMSENGINE_TESTING_UTIL_OPENSSL_ASSERTIONS_H_
+#define KMSENGINE_TESTING_UTIL_OPENSSL_ASSERTIONS_H_
+
+#include <openssl/err.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "absl/strings/string_view.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_format.h"
+
+namespace kmsengine {
+namespace testing_util {
+
+// Helper function for formatting the custom assertion message.
+inline testing::AssertionResult MakeFormattedAssertionFailure(
+    absl::string_view message,
+    absl::string_view expected,
+    absl::string_view actual) {
+  // Indentation for "Expected" and "Actual" follows googlemock ASSERT_* and
+  // EXPECT_* formatting.
+  return testing::AssertionFailure() << absl::StrFormat(
+      "%s\n"
+      "         Expected: %s\n"
+      "           Actual: %s", message, expected, actual);
+}
+
+// Returns `testing::AssertionSuccess` if `openssl_return` is true. Otherwise,
+// returns `testing::AssertionFailure` with a formatted message containing the
+// first OpenSSL error in the OpenSSL error queue.
+//
+// Defined as a predicate instead of a matcher since matchers must be purely
+// functional, but `ERR_get_error_line_data` has OpenSSL-internal side-effects.
+testing::AssertionResult IsOpenSslSuccess(const char* expr,
+                                          bool openssl_return) {
+  if (openssl_return) return testing::AssertionSuccess();
+
+  const char *file, *data;
+  int line, flags;
+  auto err = ERR_get_error_line_data(&file, &line, &data, &flags);
+
+  // If err == 0, then no error was available in the OpenSSL error queue.
+  if (err) {
+    auto message = absl::StrFormat("%s unexpectedly returned false and "
+                                   "produced OpenSSL error...", expr);
+    auto actual = absl::StrFormat("false, with error in %s:%d [%s]", file, line,
+                                  data);
+    return MakeFormattedAssertionFailure(message, "success", actual);
+  } else {
+    auto message = absl::StrFormat("%s unexpectedly returned false...", expr);
+    return MakeFormattedAssertionFailure(message, "success",
+                                         "false (no OpenSSL error)");
+  }
+}
+
+// Returns `testing::AssertionSuccess` if `openssl_return` is false and the
+// first OpenSSL error in the OpenSSL error queue contains the substring
+// `substr_error`. Otherwise, returns `testing::AssertionFailure`.
+//
+// Defined as a predicate instead of a matcher since matchers must be purely
+// functional, but `ERR_get_error_line_data` has OpenSSL-internal side-effects.
+testing::AssertionResult IsOpenSslFailure(const char* return_expr,
+                                          const char* status_expr,
```

Comment:
Maybe rename this argument? Using the word "status" might make people think that it's a `Status` object.